crates/experimentation_client/src/lib.rs

```diff
@@ -69,34 +74,92 @@ impl Client {
-        let experiments: Experiments = self.get_satisfied_experiments(context).await;
+    pub async fn get_applicable_variant(
+        &self,
+        context: &Value,
+        toss: i8,
+    ) -> Result<Vec<String>, String> {
+        let experiments: Experiments =
+            self.get_satisfied_experiments(context, None).await?;
         let mut variants: Vec<String> = Vec::new();
         for exp in experiments {
             if let Some(v) =
-                self.decide_variant(exp.traffic_percentage, exp.variants, toss)
+                self.decide_variant(exp.traffic_percentage, exp.variants, toss)?
             {
                 variants.push(v.id)
             }
         }
-        variants
+        Ok(variants)
     }
 
-    pub async fn get_satisfied_experiments(&self, context: &Value) -> Experiments {
+    pub async fn get_satisfied_experiments(
+        &self,
+        context: &Value,
+        prefix: Option<Vec<String>>,
+    ) -> Result<Experiments, String> {
         let running_experiments = self.experiments.read().await;
-        running_experiments
+        let filtered_running_experiments = running_experiments
             .iter()
             .filter(|(_, exp)| {
                 jsonlogic::apply(&exp.context, context) == Ok(Value::Bool(true))
             })
             .map(|(_, exp)| exp.clone())
-            .collect::<Experiments>()
+            .collect::<Experiments>();
+
+        if let Some(prefix) = prefix {
+            let prefix_list: HashSet<&str> = prefix.iter().map(|s| s.as_str()).collect();
+
+            let prefix_filtered_running_experiments: Vec<Experiment> =
+                filtered_running_experiments
+                    .into_iter()
+                    .filter_map(|experiment| {
+                        let variants: Vec<Variant> = experiment
+                            .variants
+                            .into_iter()
+                            .filter_map(|mut variant| {
+                                let overrides_map: Map<String, Value> =
+                                    serde_json::from_value(variant.overrides.clone())
+                                        .ok()?;
+                                let filtered_override: Map<String, Value> = overrides_map
+                                    .into_iter()
+                                    .filter(|(key, _)| {
+                                        prefix_list
+                                            .iter()
+                                            .any(|prefix_str| key.starts_with(prefix_str))
+                                    })
+                                    .collect();
+                                variant.overrides =
+                                    serde_json::to_value(filtered_override.clone())
+                                        .ok()?;
+                                if filtered_override.is_empty() {
```

Comment:
I think this check should be before calculating variant.overrides , we can save one operation here: 
```
        if filtered_override.is_empty() {
            return None; // Skip this variant
        }
        
        variant.overrides = serde_json::to_value(filtered_override).ok()?;
        Some(variant)
    })
    .collect();

```
Also you wouldn't need the else statement if you change the control structure.